<a href="https://colab.research.google.com/github/hzhoujoy/ST554_HW7/blob/main/JoyZhou_HW7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Home Work 7**   
Author: Huiping Zhou  
Date: 3/28/2026  

# Instruction
The purpose of this homework is to practice fitting MLR and logistic regression models (including penalized or regularized models).
We will use a ` Wine Quality` dataset from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/186/wine+quality).
- fixed_acidity
- volatile_acidity
- citric_acid
- residual_sugar
- chlorides
- free_sulfur_dioxide
- total_sulfur_dioxide
- density
- pH
- sulphates
- alcohol
- quality: score between 0 to 10
- color: red or white   
we will use `alcohol` as a target variable for fitting multiple linear regression model while using `type of wine` as the response variable for fitting logistic regression model.

# Read in and Combine Data  
We start by downloading the `winequality-red.csv` and `winequality-white.csv` datasets from the [UCI machine
learning repository site](https://archive.ics.uci.edu/dataset/186/wine+quality), saving them locally, and then uploaingd them into the notebook.

Next, we use the `pd.read_cvs()` function with relative file paths to read in the two datasets.


In [ ]:
import pandas as pd
import numpy as np
# read in winequality-red dataset
red_data = pd.read_csv("winequality-red.csv", sep=";")
#Display summary information
red_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


According to the info() output, `red_data` includes 1,599 rows and 12 variables.

In [ ]:
#check the first 5 rows of dataset
red_data.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [ ]:
#read in dataset
white_data = pd.read_csv("winequality-white.csv", sep=";")

#provides a summary of the dataset
white_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


According to the info() output, `white_data` includes 4,898 rows and 12 columns.

In [ ]:
#check the first 5 rows of dataset
white_data.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6


Then, we combine these two datasets and create a new variable that represents the type of wine (red or white).

In [ ]:
# Combine along columns
wine_df = pd.concat([red_data, white_data], ignore_index=True)
wine_df.info() #check summary information of combined dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 609.2 KB


According to the info() output, combined `wine_df` includes 6497 rows and 12 columns, indicating that the red and white wine datasets were merged properly.

We create a new variable, `type_of_wine`, to indicate whether each observation corresponds to red or white wine type in the combined dataset.

In [ ]:
#First part of the combined dataset (rows originally from red_data): label as "red"
wine_df.loc[:len(red_data)-1, "type_of_wine"] = "red"

# Remaining rows (originally from white_data): label as "white"
wine_df.loc[len(red_data):, "type_of_wine"] = "white"

In [ ]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type_of_wine          6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


We check if the new variable is created correctly by using 'value_counts() method.

In [ ]:
wine_df['type_of_wine'].value_counts()

,count
type_of_wine,
white,4898
red,1599


The `type_of_wine` variable was successfully created. The value_counts() output shows that it contains two categories(red and white), which match the wine types in the original dataset. The number of observations in each category also sums to the total number of rows in the original dataset, indicating the variable was created correctly.

We update the column names in the `wine_df` DataFrame by replacing spaces with underscores.

In [ ]:
wine_df.columns = [col.replace(" ", "_") for col in wine_df.columns]
wine_df.head()

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,alcohol,quality,type_of_wine
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,red
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,red
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,red
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red


In [ ]:
#check if there are missing values in wine_df
wine_df.isnull().sum()

,0
fixed_acidity,0
volatile_acidity,0
citric_acid,0
residual_sugar,0
chlorides,0
free_sulfur_dioxide,0
total_sulfur_dioxide,0
density,0
pH,0
sulphates,0


Based on the missing-value check, the results show that there are no missing values in any column of the dataset.

# Regression Task (`alcohol` as response)   
We first fit the regression models using `alcohol` as the response variable.

## Split the data
We use the `train_test_split()` function to divide the dataset into training and test sets. Stratified sampling is applied to ensure the training and test sets maintain the same proportion of red and white wines as the original dataset.

First, let's just read in all the functions we'll need from sklearn

In [ ]:
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression, LassoCV, Lasso

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

#Split the dataset into predictors (X) and response (y)
X_train, X_test, y_train, y_test = train_test_split(
    wine_df.drop(['alcohol', 'type_of_wine'], axis=1), #prodictors(X)
    wine_df['alcohol'],                                #response
    test_size=0.2,                                     #80/20 split
    random_state=42,
    stratify=wine_df['type_of_wine']                   #maintains wine type proportions in each split
)

The `alcohol` are used as a response varaible to see which chemical properties of wine predict alcohol content. We set up the predictors (X) by dropping the response variable `alcohol` and the categorical variable `type_of_wine` from the dataset, since `type_of_wine` is used only for stratified sampling in the train-test split and is not included as a predictor in the regression model.

## Train Models
Since the the variables in the wine_df dataset are on very different scales, we will standardize the predictors to imporve model performance and ensures that all features contribute eaually.
- We store the `means` and `stds` used for standardization, so we can standardize the test set using the same values.


In [ ]:
means = X_train.apply(np.mean, axis = 0)
means

,0
fixed_acidity,7.215836
volatile_acidity,0.338770
citric_acid,0.318378
residual_sugar,5.392101
chlorides,0.056070
free_sulfur_dioxide,30.524533
total_sulfur_dioxide,115.750337
density,0.994678
pH,3.217976
sulphates,0.531155


In [ ]:
stds = X_train.apply(np.std, axis = 0)
stds

,0
fixed_acidity,1.300954
volatile_acidity,0.164631
citric_acid,0.143717
residual_sugar,4.725284
chlorides,0.035602
free_sulfur_dioxide,17.771404
total_sulfur_dioxide,56.743248
density,0.003018
pH,0.160134
sulphates,0.147937


In [ ]:
X_train = X_train.apply(lambda x: (x-np.mean(x))/np.std(x), axis = 0)
X_train

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,quality
4110,-0.780839,-0.235499,1.194170,1.229111,0.054220,2.108751,1.749101,0.623768,-0.174702,0.465371,0.201651
3062,0.833361,-0.903661,1.194170,-0.887164,-0.142398,-1.098649,0.374488,-0.953600,-0.986520,-0.480979,0.201651
5236,1.140827,-0.235499,0.150449,-0.125305,-1.069314,-0.423407,0.198256,0.020657,-0.611835,-0.278190,1.341998
497,-0.012173,0.007469,0.011287,-0.612048,0.953048,0.701997,-0.048470,0.637023,0.637117,1.749702,-0.938696
826,0.218427,-0.417725,0.150449,-0.654374,-0.170487,-1.492540,-1.898910,0.139954,1.136697,0.735756,1.341998
...,...,...,...,...,...,...,...,...,...,...,...
634,0.525894,0.068211,-0.754109,-0.739025,0.475546,0.870807,-0.242326,0.570747,0.324879,0.330178,-0.938696
4273,-0.012173,0.068211,-0.475783,0.043997,-0.676077,-0.423407,0.074893,-0.443275,-1.798339,0.870949,1.341998
5343,-0.165906,-1.146628,-0.267039,1.969807,-0.142398,-0.085786,0.585967,1.074445,-1.610996,-0.954153,1.341998
3063,0.602761,0.614888,1.194170,0.784693,-0.704165,0.870807,0.621213,-0.688496,-0.362044,-1.765310,2.482344


In [ ]:
#now each column has mean 0
X_train.apply(np.mean, axis = 0)

,0
fixed_acidity,1.011741e-16
volatile_acidity,-6.699364e-17
citric_acid,-1.066429e-16
residual_sugar,3.691486e-17
chlorides,5.742312e-17
free_sulfur_dioxide,9.023633e-17
total_sulfur_dioxide,6.562642e-17
density,-2.437065e-14
pH,2.604549e-16
sulphates,-1.517611e-16


In [ ]:
#now each column has sd 1
X_train.apply(np.std, axis = 0)

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


### Fit 4 MLR Models
Fit four different multiple linear regression models.
- At least one should include interaction terms
- At least one should include some polynomial terms (you may want to standardize your predictors
but that is up to you)
- Use CV to select your best MLR model

Below I'll consider some basic MLR models to compare. Even though we don't need to use CV on those to determine a tuning parameter, we can use CV on those to determine the best MLR model (using the training set alone). Similarly, we'll then find a 'best' LASSO model with tuning parameter selected by CV. Then we'll compare the two 'best' models on the test set to see the overall winner!

First, we will fit a full model includsing all predictors

In [ ]:
# Full model
cv_full_model = cross_validate(
    LinearRegression(),
    X_train,
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

Second, we include only the most important wine-content features, `density and residual_sugar`, as predictors.

In [ ]:
cv_mlr1= cross_validate(
    LinearRegression(),
    X_train[['density','residual_sugar', 'fixed_acidity']],
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

We generated an interaction term between `density and residual_sugar` using `PolynomialFeatures` with interaction-only mode. The interaction was fit on the training set to avoid data leakage and then added back into the predictor matrix before running the multiple linear regression model.

We will include a interation term in the modle.
- To include interaction terms between `density and residual_sugar` , we use `PolynomialFeatures()` from `sklearn` tools to generate them.
 - First, we create interaction transformer

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
#create an interaction term transformer
interaction = PolynomialFeatures(interaction_only=True, include_bias = False)

- Generate transform X_train data

In [ ]:
# Generate the interaction features for the training data (fit and transform)
inter_train = interaction.fit_transform(X_train[["density", "fixed_acidity", "residual_sugar"]])

- Generate transform X_test

In [ ]:
# Transform the test data using the features learned from the training data
inter_test  = interaction.transform(X_test[["density", "fixed_acidity", "residual_sugar"]])

- Check the feature names

In [ ]:
# Create DataFrames with feature names
feature_names = interaction.get_feature_names_out(
    ['density', 'fixed_acidity', 'residual_sugar']
)
feature_names

array(['density', 'fixed_acidity', 'residual_sugar',
       'density fixed_acidity', 'density residual_sugar',
       'fixed_acidity residual_sugar'], dtype=object)

- keep the columns that we want for modeling

In [ ]:
import pandas as pd
#Keep only the desired predictors
cols_to_keep = ['density', 'fixed_acidity', 'residual_sugar','fixed_acidity residual_sugar']

X_train_inter = pd.DataFrame(inter_train, columns=feature_names)[cols_to_keep]

X_test_inter = pd.DataFrame(inter_test, columns=feature_names)[cols_to_keep]

- Finally, we perform 5-cross-validation with LinearRegression using using the training data with the added interaction terms (X_train_inter) to fit the model `cv_mlr2`.

In [ ]:
cv_mlr2 = cross_validate(
    LinearRegression(),
    X_train_inter,
    y_train,
    cv=5,
    scoring="neg_mean_squared_error"
)

Now, we incorporate a polynomial term (density²) into the multiple linear regression model, using `density and residual_sugar` as predictors.

- We use `PolynomialFeatures()` to create polynomial terms, including the squared term for density.

In [ ]:
poly = PolynomialFeatures(
    degree=2,
    interaction_only=False,  #allow squared terms
    include_bias=False
)

- Generate transform transing data

In [ ]:
poly_train = poly.fit_transform(X_train[["density", "residual_sugar"]])

- Generate transform test data

In [ ]:
poly_test = poly.transform(X_test[["density", "residual_sugar"]])

- Let us check the feature names

In [ ]:
poly.get_feature_names_out(["density", "residual_sugar"])

array(['density', 'residual_sugar', 'density^2', 'density residual_sugar',
       'residual_sugar^2'], dtype=object)

- keet the columns (density, residual_sugar, desnsity^2) for modeling

In [ ]:
import pandas as pd

cols_to_keep = ["density", "residual_sugar", "density^2"]

X_train_poly = pd.DataFrame(poly_train,
                            columns=poly.get_feature_names_out(["density","residual_sugar"])
                           )[cols_to_keep]

X_test_poly = pd.DataFrame(poly_test,
                           columns=poly.get_feature_names_out(["density","residual_sugar"])
                          )[cols_to_keep]

- Finally, we perform 5-cross-validation with LinearRegression using using the training data with the added polynomial terms (X_train_inter) to fit the model cv_mlr3.

In [ ]:
from sklearn.model_selection import cross_validate
from sklearn.linear_model import LinearRegression

cv_mlr3 = cross_validate(
    LinearRegression(),
    X_train_poly,
    y_train,
    cv=5,
    scoring="neg_mean_squared_error"
)

- We compute the cross‑validated RMSE for each of the four fitted multiple linear regression models to compare their predictive performance.

In [ ]:
print(
    f"{np.sqrt(-sum(cv_full_model['test_score'])):.4f}",
    f"{np.sqrt(-sum(cv_mlr1['test_score'])):.4f}",
    f"{np.sqrt(-sum(cv_mlr2['test_score'])):.4f}",
    f"{np.sqrt(-sum(cv_mlr3['test_score'])):.4f}"
)

1.2249 1.7775 1.7784 2.1324


Among the four models, `cv_full_model achieves` the lowest RMSE (1.2249), indicating the best predictive performance. The model `cv_mlr3` has the highest RMSE (2.1324), making it the poorest performer.


After selecting the full model via cross‑validation, we refit it on the entire training set so that we can evaluate its performance on the test set.

In [ ]:
mlr_best = LinearRegression().fit(X_train, y_train)
print(mlr_best.intercept_)
print(np.array(list(zip(X_train.columns, mlr_best.coef_))))

10.48958309280991
[['fixed_acidity' '0.6829423664553563']
 ['volatile_acidity' '0.2547477627675012']
 ['citric_acid' '0.06588733298114649']
 ['residual_sugar' '0.9137614164980166']
 ['chlorides' '-0.008729307303509681']
 ['free_sulfur_dioxide' '-0.002344368780750644']
 ['total_sulfur_dioxide' '-0.2287652724953513']
 ['density' '-1.7308513113897757']
 ['pH' '0.44207864606041286']
 ['sulphates' '0.19692528973537907']
 ['quality' '0.12026256923958925']]


The full multiple linear regression model shows that `density` has the strongest association with alcohol content (coefficient = -1.7381). `Residual_sugar` is the strongest positive predictor (coefficient = 0.9138), followed by `fixed_acidity` and `pH`. `Total_sulfur_dioxide` shows a moderate negative relationship, while `chlorides` and `free_sulfur_dioxide` have very small coefficients, indicating minimal influence on alcohol content.

### Fit a LASSO model
We use 5 predicors (`fixed_acidity, density, residual_sugar, total_sulfur_dioxide, and pH` ) to fit this model


- Subset X_train and X_test to **only** these 5 predictors

In [ ]:
X_train_lasso = X_train[['fixed_acidity','density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']]
X_test_lasso = X_test[['fixed_acidity','density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']]

- Fit the LASSO model on those predictors using CV

In [ ]:
lasso_mod = LassoCV(cv=5, random_state=0) \
    .fit(X_train_lasso,
         y_train)

Check the tuning parameters and CV errors

In [ ]:
np.set_printoptions(suppress = True)
fit_info = np.array(list(zip(lasso_mod.alphas_, np.mean(lasso_mod.mse_path_, axis = 1))))
fit_info[fit_info[:,1].argsort()][0:10,].round(4)

array([[0.0008, 0.3768],
       [0.0009, 0.3768],
       [0.0009, 0.3768],
       [0.001 , 0.3768],
       [0.0011, 0.3768],
       [0.0012, 0.3768],
       [0.0013, 0.3768],
       [0.0013, 0.3768],
       [0.0014, 0.3769],
       [0.0015, 0.3769]])

We extracted the cross-validated MSE corresponding to each candidate value of the LASSO tuning parameter $\alpha$. The lowest cross-validated MSE (0.3768) occurs at $\alpha$ = 0.
0008, indicating that the LASSO penalty has a samll effect on the model.

In [ ]:
np.set_printoptions(suppress = True)
print(lasso_mod.alpha_)
print(lasso_mod.intercept_)
print(np.array(list(zip(X_train_lasso.columns, np.round(lasso_mod.coef_, 4)))))

0.0008255721135897033
10.489583092809914
[['fixed_acidity' '0.7332']
 ['density' '-1.6203']
 ['residual_sugar' '0.8327']
 ['total_sulfur_dioxide' '-0.3127']
 ['pH' '0.4974']]


Among the five predictors included in the LASSO model, `density` shows the strongest negative association with alcohol content (coefficient = -1.6203). In contrast, `total_sulfur_dioxide` has the weakest negative effect (coefficient = -0.3127).

Fit that best model for comparison on the test set uisng the optimal tuning parameter (α=0.0008) selected by CV.

In [ ]:
lasso_best = Lasso(lasso_mod.alpha_) \
    .fit(X_train_lasso,
         y_train)

### Fit a Ridge Regression model
We fit a Ridge regression model using five selected predictors (`density, residual_sugar, total_sulfur_dioxide, and pH`). The tuning parameter λ was chosen through 5-fold cross-validation via RidgeCV, using a log-spaced grid of candidate values from 10⁻³ to 10³.

In [ ]:
from sklearn.linear_model import Ridge
import numpy as np
import pandas as pd
from sklearn.linear_model import RidgeCV

# Set up a range of possible lambda values
alphas = np.logspace(-3, 3, 50)

# Initialize the RidgeCV model to find the best lambda
ridge_cv_model = RidgeCV(alphas=alphas, cv=5)

# Fit the model on the training data
ridge_cv_model.fit(X_train[['fixed_acidity', 'density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']],
                   y_train)

# Print the best alpha (lambda) value
print(f"Optimal lambda: {ridge_cv_model.alpha_:.4f}")

Optimal lambda: 2.6827


Using 5-fold cross-validation, RidgeCV selected an optimal penalty parameter of λ = 2.6827. This value represents the amount of shrinkage that yields the lowest cross-validated mean squared error among the candidate λ values tested.

-  Fit the final Ridge regression model on the training data using the optimal tuning parameter (λ = 2.6827) selected by cross-validation.

In [ ]:
#Fit final Ridge model using selected lambda
ridge_best = Ridge(ridge_cv_model.alpha_) \
            .fit(X_train[['fixed_acidity', 'density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']],
                   y_train)

### Fit an Elastic Net model
We use five selected predictors (density, residual_sugar, total_sulfur_dioxide, and pH) to fit an Elastic Net model.

- Subset X_train and X_test to only these 5 predictors

In [ ]:
X_train_en = X_train[['fixed_acidity','density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']]
X_test_en = X_test[['fixed_acidity','density', 'residual_sugar', 'total_sulfur_dioxide', 'pH']]

In [ ]:
from sklearn.linear_model import ElasticNetCV
regr = ElasticNetCV(cv=5,
                    random_state=0,
                    l1_ratio = [0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96, 0.98, 0.99, 1],
                    n_alphas = 50)
regr.fit(X_train_en, y_train)

ElasticNetCV(cv=5,
             l1_ratio=[0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96, 0.98, 0.99, 1],
             n_alphas=50, random_state=0)

In [ ]:
print(regr.alpha_)

0.0008255721135897033


In [ ]:
print(regr.l1_ratio_)

1.0


Refit on full training data with best tuning parameters

In [ ]:
from sklearn.linear_model import ElasticNet
en = ElasticNet(alpha = regr.alpha_, l1_ratio = regr.l1_ratio_)
en.fit(X_train_en, y_train)

ElasticNet(alpha=np.float64(0.0008255721135897033), l1_ratio=np.float64(1.0))

In [ ]:
print(np.array(list(zip(X_train_en.columns, en.coef_))))

[['fixed_acidity' '0.7332464048096592']
 ['density' '-1.6203273617810647']
 ['residual_sugar' '0.8326721881487807']
 ['total_sulfur_dioxide' '-0.31265925795564103']
 ['pH' '0.49742278102503273']]


## Test Models
We need to standardize the test set using our training values. Then we can find the test set predictions and RMSE!

In [ ]:
#quick function to standardize based off of a supplied mean and std
def my_std_fun(x, means, stds):
    return(x-means)/stds
#loop through the columns and use the function on each
for x in X_test.columns:
    X_test[x] = my_std_fun(X_test[x], means[x], stds[x])

X_test.head()

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,quality
1739,-6.087693,-3.119267,-1.168469,-1.284079,-10.308372,-1.750945,-2.051931,-546.788406,-16.896777,-8.212435,-7.710869
2845,-5.083250,-7.177809,1.736472,-1.149720,-2.418815,-1.684452,-2.023979,-261.275759,-21.576416,-9.583216,-5.110088
5282,-5.792268,-9.022601,2.220629,-0.804867,-11.097328,-1.562548,-2.019010,-28.473140,-19.236597,7.780011,-6.410478
5822,-6.678541,1.308233,-12.304079,-1.333344,-22.931663,-1.795274,-2.060006,-759.824766,-8.707410,-5.927800,-9.011260
5237,-6.028608,-3.857184,-2.136783,-1.275122,-34.765998,-1.706616,-2.048204,-834.497304,-19.236597,2.296887,-5.110088


Now find our predictions and test set RMSE.

In [ ]:
mlr_pred = mlr_best.predict(X_test)
lasso_pred = lasso_best.predict(X_test)

print(np.sqrt(mean_squared_error(y_test, mlr_pred)), np.sqrt(mean_squared_error(y_test, lasso_pred)))